core/state_transition.go

```diff
@@ -693,42 +744,48 @@ func (st *StateTransition) innerTransitionDb() (*ExecutionResult, error) {
 
 		// we burn the token if gas is from SoulGasToken which is not backed by native;
 		// otherwise we add to the native balance
-		if st.gasFromSoul && st.evm.ChainConfig().IsOptimism() && !st.evm.ChainConfig().Optimism.IsSoulBackedByNative {
-		} else {
-			st.state.AddBalance(st.evm.Context.Coinbase, fee, tracing.BalanceIncreaseRewardTransactionFee)
+		if st.initialUsedSGTBalance != nil && st.evm.ChainConfig().IsOptimism() && !st.evm.ChainConfig().Optimism.IsSoulBackedByNative {
+			fee, _ = st.distributeGas(fee)
```

Comment:
I think refund to sender and pay to fee collector should behave differently
- For sender, we should refund gas token, and then SGT
- For fee collect, we should pay SGT, and then pay gas token

Reply:
Done in 5514fb52219d7f8316a71ec92875a4e5c2f35d76 .